src/client/core/page-unload-barrier.js

```diff
@@ -0,0 +1,130 @@
+import hammerhead from './deps/hammerhead';
+import * as eventUtils from './utils/event';
+import { EventEmitter } from './utils/service';
+import delay from './utils/delay';
+
+var Promise      = hammerhead.Promise;
+var browserUtils = hammerhead.utils.browser;
+
+
+const DEFAULT_BARRIER_TIMEOUT       = 500;
+const WAIT_FOR_UNLOAD_TIMEOUT       = 3000;
+const SHORT_WAIT_FOR_UNLOAD_TIMEOUT = 30;
+
+const WAITING_FINISHED_EVENT = 'waiting-finished';
+
+
+class PageUnloadMonitor extends EventEmitter {
+    constructor () {
+        super();
+
+        this.waitingForUnload          = false;
+        this.waitingForUnloadTimeoutId = null;
+        this.unloading                 = false;
+
+        this._handleSubmit();
+        this._handleLinkClick();
+        this._handleBeforeUnload();
+    }
+
+    _overrideFormSubmit (form) {
```

Comment:
I think we need to move this code to hammerhead and expose it as 'BEFORE_SUBMIT' event.